test/sqllogictest/column_knowledge.slt

```diff
@@ -477,8 +474,19 @@ query T multiline
 EXPLAIN WITH(arity, join_impls) SELECT * FROM t4 AS a1 LEFT JOIN t4 AS a2 USING (f1, f2) WHERE a1.f1 = 123 AND a1.f2 = 234;
 ----
 Explained Query:
-  Filter (#0 = 123) AND (#1 = 234) // { arity: 2 }
-    Get materialize.public.t4 // { arity: 2 }
+  Return // { arity: 2 }
+    Union // { arity: 2 }
+      Project (#1, #0) // { arity: 2 }
+        Map (234, 123) // { arity: 2 }
+          Negate // { arity: 0 }
+            Project () // { arity: 0 }
+              Get l0 // { arity: 2 }
+      Get l0 // { arity: 2 }
+      Get l0 // { arity: 2 }
+  With
+    cte l0 =
+      Filter (#0 = 123) AND (#1 = 234) // { arity: 2 }
+        Get materialize.public.t4 // { arity: 2 }
```

Comment:
This looks pretty terrible. :D

Reply:
Yeah, on cloud at the moment the `ON` version of the above join produces the gross output:
```
materialize=> EXPLAIN WITH(arity, join_impls) SELECT * FROM t4 AS a1 LEFT JOIN t4 AS a2 USING (f1, f2) WHERE a1.f1 = 123 AND a1.f2 = 234;
                   Optimized Plan                   
----------------------------------------------------
 Explained Query:                                  +
   Filter (#0 = 123) AND (#1 = 234) // { arity: 2 }+
     Get materialize.public.t4 // { arity: 2 }     +
                                                   +
 Source materialize.public.t4                      +
   filter=((#0 = 123) AND (#1 = 234))              +
 
(1 row)

Time: 17.522 ms
materialize=> EXPLAIN WITH(arity, join_impls) SELECT * FROM t4 AS a1 LEFT JOIN t4 AS a2 ON (a1.f1 = a2.f1 AND a1.f2 = a2.f2) WHERE a1.f1 = 123 AND a1.f2 = 234;
                     Optimized Plan                     
--------------------------------------------------------
 Explained Query:                                      +
   Return // { arity: 4 }                              +
     Union // { arity: 4 }                             +
       Map (null, null) // { arity: 4 }                +
         Union // { arity: 2 }                         +
           Project (#1, #0) // { arity: 2 }            +
             Map (234, 123) // { arity: 2 }            +
               Negate // { arity: 0 }                  +
                 Project () // { arity: 0 }            +
                   Get l0 // { arity: 2 }              +
           Get l0 // { arity: 2 }                      +
       Project (#0, #1, #0, #1) // { arity: 4 }        +
         Get l0 // { arity: 2 }                        +
   With                                                +
     cte l0 =                                          +
       Filter (#0 = 123) AND (#1 = 234) // { arity: 2 }+
         Get materialize.public.t4 // { arity: 2 }     +
                                                       +
 Source materialize.public.t4                          +
   filter=((#0 = 123) AND (#1 = 234))                  +
 
(1 row)

Time: 15.146 ms
materialize=> 
```

Reply:
I remember that the `true AND` had some magic powers from conversations with @wangandi and @asenac... :disappointed: I guess this query demonstrates one of them. I think this is a good candidate for side-by-side tracing, let me check what I can find.

Reply:
Indeed, this is an orthogonal problem that we are starting to see now because the optimization path for the more general `LEFT JOIN` lowering code is able to cover, but the optimized join plan does.

I'll open a separate issue and reference it here.